# Segmenting and Clustering Neighborhoods in Toronto - Part 1

### For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.
1. Start by creating a new Notebook for this assignment.
2. Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:

Link for scrapping a table from a website: https://towardsdatascience.com/web-scraping-html-tables-with-python-c9baba21059

In [1]:
# Import necessary libraries

import requests
import lxml.html as lh
import pandas as pd

In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

#Create a handle, page, to handle the contents of the website
page = requests.get(url)

#Store the contents of the website under doc
doc = lh.fromstring(page.content)

#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

In [3]:
#Check the length of the first 12 rows
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

This measn that there are 3 colums per row

In [4]:
# Parse the first row as our header
tr_elements = doc.xpath('//tr')

#Create empty list
col=[]
i=0

#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print ('%d:"%s"'%(i,name))
    col.append((name,[]))

1:"Postal Code
"
2:"Borough
"
3:"Neighborhood
"


#### Creating pandas dataframe
Each header is appended to a tuple along with an empty list.

In [5]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1
        

# Check the length of each column. Ideally, they should all be the same
[len(C) for (title,C) in col]

[181, 181, 181]

This shows that each column has 181 rows each

#### Instantiating the pandas dataframe

In [6]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

3.To create the above dataframe:

* The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
* Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
* More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
* If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
* Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
* In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

In [7]:
# Access the top 5 rows of the data frame 
df.head()

Postal Code\n           Borough\n               Neighborhood\n
0         M1A\n      Not assigned\n               Not assigned\n
1         M2A\n      Not assigned\n               Not assigned\n
2         M3A\n        North York\n                  Parkwoods\n
3         M4A\n        North York\n           Victoria Village\n
4         M5A\n  Downtown Toronto\n  Regent Park, Harbourfront\n

#### Rearranging and naming columns

In [8]:
df.columns = ['Borough', 'Neighbourhood','Postcode']

cols = df.columns.tolist()
cols

cols = cols[-1:] + cols[:-1]

df = df[cols]

df.head()

Postcode Borough       Neighbourhood
0               Not assigned\n   M1A\n      Not assigned\n
1               Not assigned\n   M2A\n      Not assigned\n
2                  Parkwoods\n   M3A\n        North York\n
3           Victoria Village\n   M4A\n        North York\n
4  Regent Park, Harbourfront\n   M5A\n  Downtown Toronto\n

#### Cleaning string data

In [9]:
df = df.replace('\n',' ', regex=True)
df.head()

Postcode Borough      Neighbourhood
0               Not assigned     M1A       Not assigned 
1               Not assigned     M2A       Not assigned 
2                  Parkwoods     M3A         North York 
3           Victoria Village     M4A         North York 
4  Regent Park, Harbourfront     M5A   Downtown Toronto

In [10]:
# Dropping all cells with a borough that is Not assigned
df.drop(df.index[df['Borough'] == 'Not assigned'], inplace = True)

# Reset the index and dropping the previous index
df = df.reset_index(drop=True)

df.head(10)

Postcode Borough      Neighbourhood
0                                 Not assigned     M1A       Not assigned 
1                                 Not assigned     M2A       Not assigned 
2                                    Parkwoods     M3A         North York 
3                             Victoria Village     M4A         North York 
4                    Regent Park, Harbourfront     M5A   Downtown Toronto 
5             Lawrence Manor, Lawrence Heights     M6A         North York 
6  Queen's Park, Ontario Provincial Government     M7A   Downtown Toronto 
7                                 Not assigned     M8A       Not assigned 
8      Islington Avenue, Humber Valley Village     M9A          Etobicoke 
9                               Malvern, Rouge     M1B        Scarborough

#### Combining Neighbourhoods based on similar Postcode and Borough

In [11]:
df = df.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(','.join).reset_index()
df.columns = ['Postcode','Borough','Neighbourhood']
df.head()

Postcode Borough  \
0                                                             
1                                        Agincourt     M1S    
2                           Alderwood, Long Branch     M8W    
3  Bathurst Manor, Wilson Heights, Downsview North     M3H    
4                                  Bayview Village     M2K    

            Neighbourhood  
0  Canadian postal codes   
1            Scarborough   
2              Etobicoke   
3             North York   
4             North York

In [12]:
df['Neighbourhood'] = df['Neighbourhood'].str.strip()

In [13]:
df.head()

Postcode Borough  \
0                                                             
1                                        Agincourt     M1S    
2                           Alderwood, Long Branch     M8W    
3  Bathurst Manor, Wilson Heights, Downsview North     M3H    
4                                  Bayview Village     M2K    

           Neighbourhood  
0  Canadian postal codes  
1            Scarborough  
2              Etobicoke  
3             North York  
4             North York

#### Assigning Borough values to the Neignbourhood where vlaue is "Not assigned"

In [14]:
df.loc[df['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df['Borough']

# Check the shape of the data frame
df.head()

Postcode Borough  \
0                                                             
1                                        Agincourt     M1S    
2                           Alderwood, Long Branch     M8W    
3  Bathurst Manor, Wilson Heights, Downsview North     M3H    
4                                  Bayview Village     M2K    

           Neighbourhood  
0  Canadian postal codes  
1            Scarborough  
2              Etobicoke  
3             North York  
4             North York

In [15]:
df.shape

(181, 3)

#### Save to csv

In [16]:
df.to_csv(r'df_can.csv')

***
***
***
# Segmenting and Clustering Neighborhoods in Toronto - Part 2

Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

Use the Geocoder package or the csv file to create the following dataframe:

In [17]:
""""import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')"""

'"import requests # library to handle requests\nimport pandas as pd # library for data analsysis\nimport numpy as np # library to handle data in a vectorized manner\nimport random # library for random number generation\n\n!conda install -c conda-forge geopy --yes \nfrom geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values\n\n# libraries for displaying images\nfrom IPython.display import Image \nfrom IPython.core.display import HTML \n    \n# tranforming json file into a pandas dataframe library\nfrom pandas.io.json import json_normalize\n\n!conda install -c conda-forge folium=0.5.0 --yes\nimport folium # plotting library\n\nprint(\'Folium installed\')\nprint(\'Libraries imported.\')'

In [18]:
""""!conda install -c conda-forge geocoder --yes 
import geocoder # import geocoder

# initialize your variable to None
lat_lng_coords = None

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]"""

'"!conda install -c conda-forge geocoder --yes \nimport geocoder # import geocoder\n\n# initialize your variable to None\nlat_lng_coords = None\n\n# loop until you get the coordinates\nwhile(lat_lng_coords is None):\n  g = geocoder.google(\'{}, Toronto, Ontario\'.format(postal_code))\n  lat_lng_coords = g.latlng\n\nlatitude = lat_lng_coords[0]\nlongitude = lat_lng_coords[1]'

#### Couldn't figure out the error, and the system was too slow. So I used the csv given in the link

In [19]:
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner

link = "http://cocl.us/Geospatial_data"
df1 = pd.read_csv(link)

df1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [20]:
df1.shape

(103, 3)

#### Changing the column name Postal code to Postcode to merge the two data frames together

In [21]:
df1.columns = ['Postcode','Latitude','Longitude']

cols = df1.columns.tolist()
cols

['Postcode', 'Latitude', 'Longitude']

#### Merging the two dataframes based on their PostCode

In [22]:
link = "https://raw.githubusercontent.com/Shekhar-rv/Coursera_Capstone/master/df_can.csv"
df = pd.read_csv(link,index_col=0)
df_new = pd.merge(df, df1, on='Postcode')
df_new.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge ,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union  43.784535   
2      M1E  Scarborough       Guildwood ,Morningside ,West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

#### Save the file as CSV

In [23]:
df_new.to_csv(r'df_final.csv')

In [24]:
df_new.shape

(103, 5)

In [25]:
#For simplifying naming conventions
neighborhoods = df_new

#### The final DataFrame has 103 rows and 5 columns

***
___
***

# Segmenting and Clustering Neighborhoods in Toronto - Part 3

#### Find how many boroughs and neighborhoods

In [26]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


#### Use geopy library to get the latitude and longitude values of Toronto, Canada

In [27]:
import json # library to handle JSON files
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Libraries imported.


In [28]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="can_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto to visualize the neighbourhoods

In [29]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Utilizing the Foursquare API to explore the neighborhoods and segment them.

In [30]:
# Define Foursquare Credentials and Version
LIMIT = 100

CLIENT_ID = 'R01LINGO2WC45KLRLKT3ZHU2QENAO2IPRK2N2ELOHRNK4P3K' 
CLIENT_SECRET = '4JT1TWRMXMPLX5IOKNBAFU3L3ARXK4D5JJDPFK1CLRZM2ZVW' 
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: R01LINGO2WC45KLRLKT3ZHU2QENAO2IPRK2N2ELOHRNK4P3K
CLIENT_SECRET:4JT1TWRMXMPLX5IOKNBAFU3L3ARXK4D5JJDPFK1CLRZM2ZVW


### Explore Neighborhoods in Toronto


#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [31]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]["groups"][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called toronto_venues.

In [32]:
toronto_venues = getNearbyVenues(names=neighborhoods['Neighbourhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude'],
                                  )

Rouge ,Malvern
Highland Creek ,Rouge Hill ,Port Union
Guildwood ,Morningside ,West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park ,Ionview ,Kennedy Park
Clairlea ,Golden Mile ,Oakridge
Cliffcrest ,Cliffside ,Scarborough Village West
Birch Cliff ,Cliffside West
Dorset Park ,Scarborough Town Centre ,Wexford Heights
Maryvale ,Wexford
Agincourt
Clarks Corners ,Sullivan ,Tam O'Shanter
Agincourt North ,L'Amoreaux East ,Milliken ,Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview ,Henry Farm ,Oriole
Bayview Village
Silver Hills ,York Mills
Newtonbrook ,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park ,Don Mills South
Bathurst Manor ,Downsview North ,Wilson Heights
Northwood Park ,York University
CFB Toronto ,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens ,Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West ,

#### Glimspe of the new data set

In [33]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                          Rouge ,Malvern              43.806686   
1  Highland Creek ,Rouge Hill ,Port Union              43.784535   
2  Highland Creek ,Rouge Hill ,Port Union              43.784535   
3       Guildwood ,Morningside ,West Hill              43.763573   
4       Guildwood ,Morningside ,West Hill              43.763573   

   Neighborhood Longitude                        Venue  Venue Latitude  \
0              -79.194353                      Wendy’s       43.807448   
1              -79.160497  Great Shine Window Cleaning       43.783145   
2              -79.160497        Royal Canadian Legion       43.782533   
3              -79.188711               RBC Royal Bank       43.766790   
4              -79.188711            G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.157431          Home Service  
2       -79.163085                   Bar  
3       -79.191151                  Bank  
4       -79.191537     Electronics Store

In [34]:
toronto_venues.shape

(2132, 7)

#### Let's check how many venues were returned for each neighborhood

In [35]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide ,King ,Richmond                                               94   
Agincourt                                                               4   
Agincourt North ,L'Amoreaux East ,Milliken ,Ste...                      2   
Albion Gardens ,Beaumond Heights ,Humbergate ,J...                      9   
Alderwood ,Long Branch                                                  7   
Bathurst Manor ,Downsview North ,Wilson Heights                        23   
Bayview Village                                                         4   
Bedford Park ,Lawrence Manor East                                      23   
Berczy Park                                                            58   
Birch Cliff ,Cliffside West                                             4   
Bloordale Gardens ,Eringate ,Markland Wood ,Old...                      8   
Brockton ,Exhibition Place ,Parkdale Village                           22   
Business Reply Mail Processing Centre 969 Eastern                      18   
CFB Toronto ,Downsview East                                             4   
CN Tower ,Bathurst Quay ,Island airport ,Harbou...                     14   
Cabbagetown ,St. James Town                                            48   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  13   
Cedarbrae                                                               8   
Central Bay Street                                                     65   
Chinatown ,Grange Park ,Kensington Market                              63   
Christie                                                               17   
Church and Wellesley                                                   76   
Clairlea ,Golden Mile ,Oakridge                                         8   
Clarks Corners ,Sullivan ,Tam O'Shanter                                14   
Cliffcrest ,Cliffside ,Scarborough Village West                         2   
Cloverdale ,Islington ,Martin Grove ,Princess G...                      1   
Commerce Court ,Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
...                                                                   ...   
Northwest                                                               2   
Northwood Park ,York University                                         6   
Parkdale ,Roncesvalles                                                 15   
Parkwoods                                                               3   
Queen's Park                                                           32   
Rosedale                                                                4   
Roselawn                                                                3   
Rouge ,Malvern                                                          1   
Runnymede ,Swansea                                                     34   
Ryerson ,Garden District                                              100   
Scarborough Village                                                     2   
St. James Town                                                         79   
Stn A PO Boxes 25 The Esplanade                                        97   
Studio District                                                        40   
The Annex ,North Midtown ,Yorkville                                    20   
The Beaches                                                             4   
The Beaches West ,India Bazaar                                         20   
The Danforth West ,Riverdale                                           43   
The Junction North ,Runnymede                                           2   
The Kingsway ,Montgomery Road ,Old Mill North   

#### Let's find out how many unique categories can be curated from all the returned venues

In [36]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 270 uniques categories.


## Analyze Each Neighborhood

In [37]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Trail  Train Station  \
0                 0                    0  ...      0              0   
1                 0                    0  ...      0              0   
2                 0                    0  ...      0              0   
3                 0                    0  ...      0              0   
4                 0                    0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  
0                0         0          0            0              0  
1                0         0          0            0              0  
2                0         0          0            0              0  
3                0         0          0            0              0  
4                0         0          0            0              0  

[5 rows x 270 columns]

In [38]:
# Examine the new data frame
toronto_onehot.shape

(2132, 270)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [39]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide ,King ,Richmond     0.000000   
1                                           Agincourt     0.000000   
2   Agincourt North ,L'Amoreaux East ,Milliken ,St...     0.000000   
3   Albion Gardens ,Beaumond Heights ,Humbergate ,...     0.000000   
4                              Alderwood ,Long Branch     0.000000   
5     Bathurst Manor ,Downsview North ,Wilson Heights     0.000000   
6                                     Bayview Village     0.000000   
7                   Bedford Park ,Lawrence Manor East     0.000000   
8                                         Berczy Park     0.000000   
9                         Birch Cliff ,Cliffside West     0.000000   
10  Bloordale Gardens ,Eringate ,Markland Wood ,Ol...     0.000000   
11       Brockton ,Exhibition Place ,Parkdale Village     0.000000   
12  Business Reply Mail Processing Centre 969 Eastern     0.000000   
13                        CFB Toronto ,Downsview East     0.000000   
14  CN Tower ,Bathurst Quay ,Island airport ,Harbo...     0.000000   
15                        Cabbagetown ,St. James Town     0.000000   
16                                Caledonia-Fairbanks     0.000000   
17              Canada Post Gateway Processing Centre     0.000000   
18                                          Cedarbrae     0.000000   
19                                 Central Bay Street     0.015385   
20          Chinatown ,Grange Park ,Kensington Market     0.000000   
21                                           Christie     0.000000   
22                               Church and Wellesley     0.026316   
23                    Clairlea ,Golden Mile ,Oakridge     0.000000   
24            Clarks Corners ,Sullivan ,Tam O'Shanter     0.000000   
25    Cliffcrest ,Cliffside ,Scarborough Village West     0.000000   
26  Cloverdale ,Islington ,Martin Grove ,Princess ...     0.000000   
27                     Commerce Court ,Victoria Hotel     0.000000   
28                                         Davisville     0.000000   
29                                   Davisville North     0.000000   
..                                                ...          ...   
69                                          Northwest     0.000000   
70                    Northwood Park ,York University     0.000000   
71                             Parkdale ,Roncesvalles     0.000000   
72                                          Parkwoods     0.000000   
73                                       Queen's Park     0.031250   
74                                           Rosedale     0.000000   
75                                           Roselawn     0.000000   
76                                     Rouge ,Malvern     0.000000   
77                                 Runnymede ,Swansea     0.029412   
78                           Ryerson ,Garden District     0.000000   
79                                Scarborough Village     0.000000   
80                                     St. James Town     0.000000   
81                    Stn A PO Boxes 25 The Esplanade     0.010309   
82                                    Studio District     0.025000   
83                The Annex ,North Midtown ,Yorkville     0.000000   
84                                        The Beaches     0.000000   
85                     The Beaches West ,India Bazaar     0.000000   
86                       The Danforth West ,Riverdale     0.023256   
87                      The Junction North ,Runnymede     0.000000   
88      The Kingsway ,Montgomery Road ,Old Mill North     0.000000   
89                                   Thorncliffe Park     0.050000   
90                                   Victoria Village     0.000000   
91                                          Westmount     0.000000   
92                                             Weston     0.000000   
93                                   Willowdale South     0.000000   
94                                    Willowdale We

In [40]:
# confirm the new size
toronto_grouped.shape

(99, 270)

#### Let's print each neighborhood along with the top 5 most common venues

In [41]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide ,King ,Richmond----
             venue  freq
0      Coffee Shop  0.10
1             Café  0.05
2              Gym  0.04
3       Restaurant  0.04
4  Thai Restaurant  0.03


----Agincourt----
                       venue  freq
0  Latin American Restaurant  0.25
1               Skating Rink  0.25
2             Breakfast Spot  0.25
3                     Lounge  0.25
4                Yoga Studio  0.00


----Agincourt North ,L'Amoreaux East ,Milliken ,Steeles East----
                      venue  freq
0                      Park   0.5
1                Playground   0.5
2  Mediterranean Restaurant   0.0
3         Mobile Phone Shop   0.0
4        Miscellaneous Shop   0.0


----Albion Gardens ,Beaumond Heights ,Humbergate ,Jamestown ,Mount Olive ,Silverstone ,South Steeles ,Thistletown----
                  venue  freq
0         Grocery Store  0.22
1  Fast Food Restaurant  0.11
2              Pharmacy  0.11
3   Fried Chicken Joint  0.11
4        Sandwich Place  0.11


----Alderwood 

              venue  freq
0  Department Store   0.2
1       Coffee Shop   0.2
2       Bus Station   0.2
3    Discount Store   0.2
4     Train Station   0.2


----East Toronto----
               venue  freq
0               Park  0.25
1        Coffee Shop  0.25
2  Convenience Store  0.25
3      Metro Station  0.25
4        Men's Store  0.00


----Emery ,Humberlea----
                           venue  freq
0  Paper / Office Supplies Store  0.33
1                 Baseball Field  0.33
2                    Fabric Shop  0.33
3                Organic Grocery  0.00
4              Mobile Phone Shop  0.00


----Fairview ,Henry Farm ,Oriole----
                  venue  freq
0        Clothing Store  0.09
1           Coffee Shop  0.08
2  Fast Food Restaurant  0.06
3     Mobile Phone Shop  0.03
4            Food Court  0.03


----First Canadian Place ,Underground city----
         venue  freq
0  Coffee Shop  0.10
1         Café  0.08
2        Hotel  0.05
3   Restaurant  0.05
4          Gym  0.04


--

            venue  freq
0  Sandwich Place  0.15
1            Café  0.15
2     Coffee Shop  0.10
3        Pharmacy  0.05
4    Liquor Store  0.05


----The Beaches----
               venue  freq
0              Trail  0.25
1  Health Food Store  0.25
2                Pub  0.25
3        Yoga Studio  0.00
4  Mobile Phone Shop  0.00


----The Beaches West ,India Bazaar----
                  venue  freq
0        Sandwich Place  0.10
1  Fast Food Restaurant  0.10
2                   Gym  0.05
3           Pizza Place  0.05
4            Board Shop  0.05


----The Danforth West ,Riverdale----
                venue  freq
0    Greek Restaurant  0.19
1  Italian Restaurant  0.07
2         Coffee Shop  0.07
3           Bookstore  0.05
4      Ice Cream Shop  0.05


----The Junction North ,Runnymede----
                      venue  freq
0               Pizza Place   0.5
1         Convenience Store   0.5
2               Yoga Studio   0.0
3  Mediterranean Restaurant   0.0
4         Mobile Phone Shop   0.0


#### Let's put that into a pandas dataframe


First, let's write a function to sort the venues in descending order.


In [42]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.



In [43]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0                           Adelaide ,King ,Richmond   
1                                          Agincourt   
2  Agincourt North ,L'Amoreaux East ,Milliken ,St...   
3  Albion Gardens ,Beaumond Heights ,Humbergate ,...   
4                             Alderwood ,Long Branch   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                Coffee Shop                  Café                   Gym   
1  Latin American Restaurant                Lounge        Breakfast Spot   
2                       Park            Playground         Women's Store   
3              Grocery Store              Pharmacy   Fried Chicken Joint   
4                Pizza Place                   Gym        Sandwich Place   

  4th Most Common Venue        5th Most Common Venue 6th Most Common Venue  \
0            Restaurant                        Hotel         Deli / Bodega   
1          Skating Rink  Eastern European Restaurant   Dumpling Restaurant   
2   Distribution Center                Deli / Bodega      Department Store   
3        Discount Store               Sandwich Place           Pizza Place   
4              Pharmacy                          Pub           Coffee Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0        Thai Restaurant             Bookstore           Pizza Place   
1              Drugstore            Donut Shop  Ethiopian Restaurant   
2           Dessert Shop    Dim Sum Restaurant                 Diner   
3   Fast Food Restaurant            Beer Store               Dog Run   
4  General Entertainment               Gay Bar               Dog Run   

  10th Most Common Venue  
0            Salad Place  
1       Doner Restaurant  
2         Discount Store  
3       Doner Restaurant  
4    Distribution Center

## Cluster Neighborhoods
Run k-means to cluster the neighborhood into 5 clusters.

In [44]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 0, 2, 2, 1, 2, 1, 1, 1], dtype=int32)

#### Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.



In [45]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge ,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union  43.784535   
2      M1E  Scarborough       Guildwood ,Morningside ,West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             2.0  Fast Food Restaurant         Women's Store   
1 -79.160497             1.0          Home Service                   Bar   
2 -79.188711             1.0    Mexican Restaurant         Moving Target   
3 -79.216917             1.0           Coffee Shop     Korean Restaurant   
4 -79.239476             1.0           Gas Station   Fried Chicken Joint   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0               Dog Run      Department Store          Dessert Shop   
1         Women's Store         Deli / Bodega          Dessert Shop   
2        Medical Center                  Bank          Intersection   
3         Women's Store               Dog Run          Dessert Shop   
4      Hakka Restaurant                Bakery    Athletics & Sports   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Dim Sum Restaurant                 Diner        Discount Store   
1    Dim Sum Restaurant                 Diner        Discount Store   
2   Rental Car Location     Electronics Store        Breakfast Spot   
3    Dim Sum Restaurant                 Diner        Discount Store   
4                  Bank  Caribbean Restaurant       Thai Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0   Distribution Center       Doner Restaurant  
1   Distribution Center                Dog Run  
2               Dog Run       Doner Restaurant  
3   Distribution Center             Donut Shop  
4                 Diner           Dessert Shop

#### Finally, let's visualize the resulting clusters



In [46]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        fill=True,
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Clusters
Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster.

## Cluster 1

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5        Scarborough             0.0         Grocery Store   
14       Scarborough             0.0                  Park   
48   Central Toronto             0.0          Tennis Court   
50  Downtown Toronto             0.0                  Park   
74              York             0.0                  Park   
90         Etobicoke             0.0                 River   
98              York             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5             Playground               Dog Run      Department Store   
14            Playground         Women's Store   Distribution Center   
48            Playground         Women's Store        Discount Store   
50                 Trail            Playground   Distribution Center   
74         Women's Store                  Pool         Deli / Bodega   
90                  Park   Distribution Center         Deli / Bodega   
98         Women's Store               Dog Run      Department Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5           Dessert Shop    Dim Sum Restaurant                 Diner   
14         Deli / Bodega      Department Store          Dessert Shop   
48         Deli / Bodega      Department Store          Dessert Shop   
50         Deli / Bodega      Department Store          Dessert Shop   
74      Department Store          Dessert Shop    Dim Sum Restaurant   
90      Department Store          Dessert Shop    Dim Sum Restaurant   
98          Dessert Shop    Dim Sum Restaurant                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5         Discount Store   Distribution Center       Doner Restaurant  
14    Dim Sum Restaurant                 Diner         Discount Store  
48    Dim Sum Restaurant                 Diner    Distribution Center  
50    Dim Sum Restaurant                 Diner         Discount Store  
74                 Diner        Discount Store    Distribution Center  
90                 Diner        Discount Store          Women's Store  
98        Discount Store   Distribution Center       Doner Restaurant

## Cluster 2

In [48]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
1         Scarborough             1.0                   Home Service   
2         Scarborough             1.0             Mexican Restaurant   
3         Scarborough             1.0                    Coffee Shop   
4         Scarborough             1.0                    Gas Station   
6         Scarborough             1.0               Department Store   
7         Scarborough             1.0                       Bus Line   
8         Scarborough             1.0                          Motel   
9         Scarborough             1.0                College Stadium   
11        Scarborough             1.0                    Auto Garage   
12        Scarborough             1.0      Latin American Restaurant   
17         North York             1.0             Athletics & Sports   
18         North York             1.0                 Clothing Store   
22         North York             1.0               Ramen Restaurant   
23         North York             1.0     Construction & Landscaping   
25         North York             1.0     Construction & Landscaping   
26         North York             1.0            Japanese Restaurant   
27         North York             1.0                            Gym   
28         North York             1.0                           Bank   
29         North York             1.0             Miscellaneous Shop   
30         North York             1.0                    Snack Place   
32         North York             1.0               Business Service   
34         North York             1.0              French Restaurant   
36          East York             1.0                   Skating Rink   
37       East Toronto             1.0                          Trail   
38          East York             1.0                    Coffee Shop   
40          East York             1.0              Convenience Store   
41       East Toronto             1.0               Greek Restaurant   
42       East Toronto             1.0                 Sandwich Place   
43       East Toronto             1.0                           Café   
44    Central Toronto             1.0                       Bus Line   
..                ...             ...                            ...   
62         North York             1.0                 Sandwich Place   
63    Central Toronto             1.0                 Ice Cream Shop   
64    Central Toronto             1.0                          Trail   
65    Central Toronto             1.0                           Café   
66   Downtown Toronto             1.0                           Café   
67   Downtown Toronto             1.0                           Café   
68   Downtown Toronto             1.0                Airport Service   
69   Downtown Toronto             1.0                    Coffee Shop   
70   Downtown Toronto             1.0                    Coffee Shop   
71         North York             1.0                 Clothing Store   
72         North York             1.0               Sushi Restaurant   
73               York             1.0                          Trail   
75   Downtown Toronto             1.0                  Grocery Store   
76       West Toronto             1.0                       Pharmacy   
77       West Toronto             1.0                            Bar   
78       West Toronto             1.0                           Café   
79         North York             1.0               Basketball Court   
80               York             1.0                    Coffee Shop   
82       West Toronto             1.0                           Café   
83       West Toronto             1.0                      Gift Shop   
84       West Toronto             1.0                           Café   
85       Queen's Park             1.0                    Coffee Shop   
86        Mississauga             1.0                          Hotel   
87       East Toronto             1.0             Ligh

## Cluster 3

In [49]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0    Scarborough             2.0  Fast Food Restaurant         Women's Store   
10   Scarborough             2.0     Indian Restaurant    Light Rail Station   
13   Scarborough             2.0           Pizza Place              Pharmacy   
15   Scarborough             2.0    Chinese Restaurant              Pharmacy   
19    North York             2.0    Chinese Restaurant                  Café   
24    North York             2.0         Grocery Store              Pharmacy   
31    North York             2.0         Grocery Store                  Park   
33    North York             2.0         Grocery Store        Discount Store   
35     East York             2.0           Pizza Place  Fast Food Restaurant   
39     East York             2.0     Indian Restaurant                   Gym   
89     Etobicoke             2.0           Pizza Place                   Gym   
99     Etobicoke             2.0           Pizza Place        Discount Store   
100    Etobicoke             2.0     Mobile Phone Shop                  Park   
101    Etobicoke             2.0         Grocery Store              Pharmacy   

     3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Dog Run      Department Store          Dessert Shop   
10   Vietnamese Restaurant             Pet Store    Chinese Restaurant   
13            Noodle House                  Bank   Fried Chicken Joint   
15           Grocery Store          Noodle House           Coffee Shop   
19                    Bank   Japanese Restaurant         Women's Store   
24                    Bank        Discount Store           Pizza Place   
31           Shopping Mall                  Bank               Dog Run   
33            Liquor Store    Athletics & Sports  Gym / Fitness Center   
35                    Bank    Athletics & Sports          Intersection   
39              Restaurant           Pizza Place              Pharmacy   
89          Sandwich Place              Pharmacy                   Pub   
99            Intersection        Sandwich Place    Chinese Restaurant   
100            Pizza Place        Sandwich Place                 Diner   
101    Fried Chicken Joint        Discount Store        Sandwich Place   

         6th Most Common Venue      7th Most Common Venue  \
0           Dim Sum Restaurant                      Diner   
10               Women's Store        Distribution Center   
13          Italian Restaurant          Convenience Store   
15                 Supermarket                       Bank   
19          Dim Sum Restaurant                      Diner   
24                     Butcher                Coffee Shop   
31                Dessert Shop         Dim Sum Restaurant   
33     Comfort Food Restaurant               Dessert Shop   
35                    Pharmacy                  Gastropub   
39                        Park  Middle Eastern Restaurant   
89                 Coffee Shop      General Entertainment   
99   Middle Eastern Restaurant                Coffee Shop   
100               Dance Studio              Deli / Bodega   
101                Pizza Place       Fast Food Restaurant   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Discount Store   Distribution Center       Doner Restaurant  
10           Dessert Shop    Dim Sum Restaurant                  Diner  
13        Thai Restaurant  Fast Food Restaurant    Rental Car Location  
15         Breakfast Spot        Sandwich Place            Pizza Place  
19         Discount Store   Distribution Center                Dog Run  
24             Donut Shop             Drugstore    Dumpling Restaurant  
31                  Diner        Discount Store    Distribution Center  
33   Ethiopian Restaurant  Colombian Restaurant      Electronics Store  
35                   Café        Breakfast Spot   Gym / Fitness Center  
39           Liquor Store         Grocery Store    

## Cluster 4

In [50]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
94  Etobicoke             3.0           Golf Course         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
94               Dog Run      Department Store          Dessert Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
94    Dim Sum Restaurant                 Diner        Discount Store   

   9th Most Common Venue 10th Most Common Venue  
94   Distribution Center       Doner Restaurant

## Cluster 5

In [51]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
81        York             4.0     Convenience Store           Pizza Place   
96  North York             4.0           Pizza Place         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
81         Women's Store      Department Store          Dessert Shop   
96               Dog Run      Department Store          Dessert Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
81    Dim Sum Restaurant                 Diner        Discount Store   
96    Dim Sum Restaurant                 Diner        Discount Store   

   9th Most Common Venue 10th Most Common Venue  
81   Distribution Center                Dog Run  
96   Distribution Center       Doner Restaurant

# Cluster 2 > Cluster 3 > Cluster 1

# Cluster 5 has two neighbours

# Cluster 4 has one neighbour